libqtile/widget/check_updates.py

```diff
@@ -56,29 +56,49 @@ def __init__(self, **config):
                          "Mandriva": ("urpmq --auto-select", 0)
                          }
 
-        # Check if distro name is valid.
-        try:
-            self.cmd = self.cmd_dict[self.distro][0].split()
-            self.subtr = self.cmd_dict[self.distro][1]
-        except KeyError:
-            distros = sorted(self.cmd_dict.keys())
-            logger.error(self.distro + ' is not a valid distro name. ' +
-                         'Use one of the list: ' + str(distros) + '.')
-            self.cmd = None
+        # Backwards compatibility for custom_command. Convert string to tuple.
+        if type(self.custom_command) is not tuple:  # Maybe remove this if statement after deprecation time period expires?
```

Comment:
Please introduce a migration for this.

Reply:
Sorry I'm new to git related stuff so I'm sure what you mean by migration. Are you referring to the 'libqtile.widget.base.deprecated' mentioned in the docs?

Reply:
We now have a command [`qtile migrate`](https://github.com/qtile/qtile/blob/master/libqtile/scripts/migrate.py) which attempts to fix a config that got broken by changes like this. You'd want to do something like the migration for `Tile` which looks like this:
```py
def tile_master_windows_rename(config):
    return (
        bowler.Query(config)
        .select_function("Tile")
        .modify_argument("masterWindows", "master_length")
    )
```
Unfortunately, we don't have an example of how to modify an argument to a tuple where the first item is the original. Hence it is time to search for this in the `bowler` documentation.
Also, I am not sure if we want to make this mandatory. For many users the default `subtr` has not been a problem (I think?) so I think a check if it is a tuple and then adding the 0 in the `__init__` would be enough.

Reply:
Okay @tych0 @yobleck. Here is a minimal partially working migration:
<details><summary><code>test_class.py</code>, approx. <code>libqtile.widget.checkupdates</code></summary>
<pre>
class CheckUpdates:
    def __init__(self, *, custom_command = (None, 0)):
        self.custom_command = custom_command
</pre>
</details>
<details><summary><code>test.py</code>, approx. <code>~/.config/qtile/config.py</code></summary>
<pre>
from test_class import CheckUpdates
CheckUpdates(custom_command='pamac checkupdates')
CheckUpdates(custom_command=' '.join(['pamac', 'checkupdates']))
CheckUpdates(custom_command=('pamac checkupdates' if HOSTNAME == 'xy' else None))
CheckUpdates(custom_command=('pamac checkupdates', 42))
</pre>
</details>

<details><summary><code>fix.py</code>, approx. <code>qtile migrate</code></summary>
<pre>
import bowler
from bowler.types import Leaf, Node, SYMBOL
from fissix.pygram import token
def my_modifier(node, capture, filename):
    if 'class_call' not in capture:
        return
    # Sadly, bowler only provides FunctionSpec, not ClassSpec.
    capture['function_name'] = capture['class_name']
    capture['function_arguments'] = capture['class_arguments']
    capture['function_parameters'] = node.children[1]
    spec = bowler.FunctionSpec.build(node, capture)
    for i, argument in enumerate(spec.arguments):
        if argument.name != 'custom_command':
            continue
        if not argument.value.children:
            new_node = Node(SYMBOL.atom, [
                Leaf(token.LPAR, '('),
                Node(SYMBOL.testlist_gexp, [
                    Leaf(argument.value.type, argument.value.value),
                    Leaf(token.COMMA, ','),
                    Leaf(token.NUMBER, '0', prefix=' ')]),
                Leaf(token.RPAR, ')')
            ])
            spec.arguments[i] = bowler.FunctionArgument(name=argument.name, value = new_node,
                                                        annotation=argument.annotation,
                                                        star=argument.star, prefix=argument.prefix)
            break
    spec.explode()
if __name__ == '__main__':
    (bowler.Query('test.py')
     .select_class('CheckUpdates')
     .modify(my_modifier)
     .execute(write=True, interactive=True))
</pre>
</details>

Although I am not a `bowler` expert and haven't used it really, I think it is waaaaaaaaaaaaay easier to maintain a short `if not isinstance(self.custom_command, (list, tuple)): ...` than this bulky migration which doesn't even always work:
```sh
$ python fix.py
--- test.py
+++ test.py
@@ -1,6 +1,6 @@
 from test_class import CheckUpdates
-CheckUpdates(custom_command='pamac checkupdates')
+CheckUpdates(custom_command=('pamac checkupdates', 0))
 CheckUpdates(custom_command=' '.join(['pamac', 'checkupdates']))
 CheckUpdates(custom_command=('pamac checkupdates' if HOSTNAME == 'xy' else None))
 CheckUpdates(custom_command=('pamac checkupdates', 42))
Apply this hunk [y,N,q,a,d,?]?
```
The second version is not caught. This is definitely a pathological case, but it would also fail if it was
`CheckUpdates(custom_command=('x' if HOSTNAME == 'MACHINE_A' else 'y'))`. This is definitely realistic as only one machine would need to run on an os that is not in the `self.cmd_dict`.

Reply:
> I think it is waaaaaaaaaaaaay easier to maintain a short if not isinstance(self.custom_command, (list, tuple)): ... than this bulky migration which doesn't even always work:

Seems like perhaps we should invent a different API for this, then. Continually asking our users to change their configs is not sustainable, and I'd like to move away from it if possible.

More generally, I don't really understand the motivation for the feature. In particular, neither this patch or the original issue have examples of problematic output, which might be useful to if we want to design alternative solutions.

Reply:
I think the check that is currently meant to be the backwards compatibility layer could just be a normal part such that the argument type becomes `custom_command: Union[str, Tuple[str, int]]` where the default 0 is added if using the `str` variant. I don't think we need to deprecate the `str` variant as this makes everything (the migration, config breakage, ...) harder.
Maybe @yobleck can add a rationale to the pr description answering your second point.

Reply:
> More generally, I don't really understand the motivation for the feature

Personally I just wanted to use pamac instead of pacman and it outputs some text even if there are no updates so subtr = 0 wasn't helpful. I later discovered a -q option that removes extra lines but this still seemed like a better idea than trying to to include every possible package manager in the cmd_dict.

> Continually asking our users to change their configs is not sustainable

Agreed. I think the string to list/tuple converter should be good enough in the short term that people won't have to update their configs if they don't want to or are unaware of the change.

Reply:
Ah I also didn't see `-q`. So the original main aspect of this pr --- changing the type of `custom_command` is rather a question to make it consistent with the `cmd_dict`. The refactoring to `__init__` also seems a good idea instead of having checks everywhere. But this really does not belong here since the original comment was about a migration for a breaking change that just makes things hard.